### ISSUE TITLE
Backward search doesn't work on neovim + Skim

### ISSUE THREAD

**Issue by AlexFreik** [lervag/vimtex#2157]
### Description

My `~/.latexmkrc'`:

```perl
$pdflatex = 'pdflatex -synctex=1';
```

My `~/.config/nvim/init.vim` (checked that the shortened version below still reproduces the problem):

```vim
call plug#begin('~/vim/plugged')
Plug 'lervag/vimtex'
    let g:vimtex_view_method='skim'
    let g:vimtex_view_skim_activate=1
    let g:vimtex_view_skim_reading_bar=1
call plug#end()

function! SetServerName()
  if has('win32')
    let nvim_server_file = $TEMP . "/curnvimserver.txt"
  else
    let nvim_server_file = "/tmp/curnvimserver.txt"
  endif
  let cmd = printf("echo %s > %s", v:servername, nvim_server_file)
  call system(cmd)
endfunction

augroup vimtex_common
    autocmd!
    autocmd FileType tex call SetServerName()
augroup END
```

## How I fixed the problem
However, backward search works when using the solution [from this article](https://jdhao.github.io/2021/02/20/inverse_search_setup_neovim_vimtex/).

TL;DR:

1. Add the following setting to nvim config.
    ```vim
    function! SetServerName()
      if has('win32')
        let nvim_server_file = $TEMP . "/curnvimserver.txt"
      else
        let nvim_server_file = "/tmp/curnvimserver.txt"
      endif
      let cmd = printf("echo %s > %s", v:servername, nvim_server_file)
      call system(cmd)
    endfunction

    augroup vimtex_common
        autocmd!
        autocmd FileType tex call SetServerName()
    augroup END
    ```

2. In the Skim Preferences menu, select "Sync > Custom" and set

    ``Args: --servername `cat /tmp/curnvimserver.txt` +"%line" "%file"``

    (instead of `Args: --remote +"%line" "%file"`)

### Steps to reproduce

1. Follow the instructions from `:help vimtex-faq-skimviewer`
2. Try to use Backward search (with Shift-Cmd-Click)

### Expected behavior

Backward search

### Actual behavior

nothing

### Do you use a latexmkrc file?

Yes

### VimtexInfo

```yaml
System info
  OS: macOS 11.4 (20F71)
  Vim version: NVIM v0.6.0-dev+170-gdb1b0ee3b
  Has clientserver: true
  Servername: /var/folders/b6/2r1sc5kn7rz3mb9kqv9439gr0000gn/T/nvim79CnEo/0

VimTeX project: homework1_answersheet
  base: homework1_answersheet.tex
  root: /Users/alexfreik/Documents/GitHub/TUD-CSE/hw1
  tex: /Users/alexfreik/Documents/GitHub/TUD-CSE/hw1/homework1_answersheet.tex
  out: /Users/alexfreik/Documents/GitHub/TUD-CSE/hw1/homework1_answersheet.pdf
  log: /Users/alexfreik/Documents/GitHub/TUD-CSE/hw1/homework1_answersheet.log
  aux: /Users/alexfreik/Documents/GitHub/TUD-CSE/hw1/homework1_answersheet.aux
  fls: /Users/alexfreik/Documents/GitHub/TUD-CSE/hw1/homework1_answersheet.fls
  main parser: current file verified
  source files:
    homework1_answersheet.tex
    header.tex
  compiler: latexmk
    configuration: 
      continuous: 1
      callback: 1
      latexmk options:
        -verbose
        -file-line-error
        -synctex=1
        -interaction=nonstopmode
      latexmk engine: -pdf
    job: 3
    pid: 26882
    cmd: max_print_line=2000 latexmk -verbose -file-line-error -synctex=1 -interaction=nonstopmode -pdf -pvc -view=none -e '$success_cmd = ($success_cmd ? $success_cmd . " ; " : "") . "echo vimtex_compiler_callback_success"' -e '$failure_cmd = ($failure_cmd ? $failure_cmd . " ; " : "") . "echo vimtex_compiler_callback_failure"' 'homework1_answersheet.tex'
  viewer: Skim
    process: 
      pid: -
      cmd: osascript -e 'set theFile to POSIX file "/Users/alexfreik/Documents/GitHub/TUD-CSE/hw1/homework1_answersheet.pdf"' -e 'set thePath to POSIX path of (theFile as alias)' -e 'tell application "Skim"' -e 'try' -e 'set theDocs to get documents whose path is thePath' -e 'if (count of theDocs) > 0 then revert theDocs' -e 'end try' -e 'open theFile' -e 'end tell' >/dev/null 2>&1 &
  qf: LaTeX logfile
  document class: exam
  packages: amsbsy amsfonts amsgen amsmath amsopn amssymb amstext amsthm array atbegshi atbegshi-ltx atveryend atveryend-ltx auxhook babel bigintcalc bitset booktabs color datetime datetime-defaults enumerate epstopdf-base etexcmds etoolbox fcnumparser fcprefix fmtcount gettitlestring graphics graphicx hycolor hyperref iftex ifthen infwarerr intcalc keyval kvdefinekeys kvoptions kvsetkeys letltxmacro ltxcmds mathabx mathdots mathrsfs nameref pdfescape pdftexcmds pgf pgfcomp-version-0-65 pgfcomp-version-1-18 pgfcore pgffor pgfkeys pgfmath pgfplots pgfrcs pgfsys refcount rerunfilecheck tabularx tikz trig uniquecounter url xcolor xkeyval
```

**Comment by clason** [comment]
Your config is not minimal ;) What happens when you remove the autocommand (and the `SetServerName` function)?

**Comment by AlexFreik** [comment]
> Your config is not minimal ;) What happens when you remove the autocommand (and the `SetServerName` function)?

It is still not working (and my current workaround also stops working, of course)

**Comment by clason** [comment]
Weird. How did you install neovim (and what version)?

**Comment by AlexFreik** [comment]
> Weird. How did you install neovim (and what version)?

brew install neovim --HEAD

**Comment by lervag** [comment]
Notice that you are not on the latest VimTeX. It may not be crucial here, but it's best practise to ensure you're on the latest commits.

**Comment by AlexFreik** [comment]
> Notice that you are not on the latest VimTeX. It may not be crucial here, but it's best practise to ensure you're on the latest commits.

Thanks for advice! Updated VimTex, but still the same problem)

**Comment by lervag** [comment]
Can you simplify even more, that is, use the following minimal `test.vim`:

```vim
set nocompatible
let &rtp = '~/.vim/plugged/vimtex,' . &rtp
let &rtp .= ',~/.vim/plugged/vimtex/after'
filetype plugin indent on
syntax enable

let g:vimtex_view_method = 'skim'
```

Also, could you for now remove the `.latexmkrc` file (I'm 99.9% sure it's not relevant, but still).

With that, you can open your TeX file with `nvim -u test.vim main.tex` and again reproduce your issue.

Also, please: Can you be very specific/verbose in exactly the steps you use to reproduce? Your current description is _not_ specific. This would be an example (but it may not be exactly what you are doing):

1. `nvim -u test.vim main.tex`
2. Start compilation: `\ll`
3. Observe that Skim is opened after successful compilation
4. Now do "Shift-Cmd-Click" within Skim

Expect: Backward search

Observe: Nothing

**Comment by AlexFreik** [comment]
> Can you simplify even more, that is, use the following minimal `test.vim`:
> 
> ```viml
> set nocompatible
> let &rtp = '~/.vim/plugged/vimtex,' . &rtp
> let &rtp .= ',~/.vim/plugged/vimtex/after'
> filetype plugin indent on
> syntax enable
> 
> let g:vimtex_view_method = 'skim'
> ```
> 
> Also, could you for now remove the `.latexmkrc` file (I'm 99.9% sure it's not relevant, but still).
> 
> With that, you can open your TeX file with `nvim -u test.vim main.tex` and again reproduce your issue.
> 
> Also, please: Can you be very specific/verbose in exactly the steps you use to reproduce? Your current description is _not_ specific. This would be an example (but it may not be exactly what you are doing):
> 
> 1. `nvim -u test.vim main.tex`
> 2. Start compilation: `\ll`
> 3. Observe that Skim is opened after successful compilation
> 4. Now do "Shift-Cmd-Click" within Skim
> 
> Expect: Backward search
> 
> Observe: Nothing

I have done exactly the same you have written. (created test.vim, deleted .latexmkrc, and reproduce the steps you wrote) And still the same. Mystery.....

**Comment by clason** [comment]
Now that we are grasping at straws (and since you specify it): you _did_ set the `Command` in the Skim Custom settings to `nvr`?

**Comment by AlexFreik** [comment]
> Now that we are grasping at straws (and since you didn't specify it): you _did_ set the `Command` in the Skim Custom settings to `nvr`?

<img width="504" alt="Screenshot 2021-09-04 at 01 13 34" src="https://user-images.githubusercontent.com/61039123/132073350-7f2fa637-1d98-4e88-8242-d87ef4fb776f.png">

**Comment by AlexFreik** [comment]
> Now that we are grasping at straws (and since you didn't specify it): you _did_ set the `Command` in the Skim Custom settings to `nvr`?

I think we can wait if anyone else have this problem. If I am the only one it wouldn't matter) Thanks for your time!)

**Comment by lervag** [comment]
Well, this is all very strange. As I don't use and don't know MacOS in any depth, it is hard to be of any more help; sorry. Of course, if you have a method for getting around this issue (as indicated in the original post), then that's very good. If it turns out that this is a problem for more people, then it might be useful/necessary to update the docs here.

**Comment by trinh** [comment]
@AlexFreik I can confirm that the method posted by jdhao and posted [here](https://jdhao.github.io/2021/02/20/inverse_search_setup_neovim_vimtex/) works while the one that I tried via the vimtex documentation does not. 

@lervag What is the best way to test the nvr command without using Skim? 

I don't really understand how 

`nvr --remote ...`

is supposed to find the server name initiated by the nvim host?

Does it matter which terminal we use or how we launch nvim? In any case I have tested jdhao's solution using iTerm and Alacritty.

**Comment by ThSGM** [comment]
@AlexFreik I can confirm that the method posted by jdhao and posted [here](https://jdhao.github.io/2021/02/20/inverse_search_setup_neovim_vimtex/) works while the one that I tried via the vimtex documentation does not. 

@lervag What is the best way to test the nvr command without using Skim? 

I don't really understand how 

`nvr --remote ...`

is supposed to find the server name initiated by the nvim host?

Does it matter which terminal we use or how we launch nvim? In any case I have tested jdhao's solution using iTerm and Alacritty.

**Comment by lervag** [comment]
@ThSGM `nvr` does not search for a server. It either uses the server specified (`nvr --servername ...`), or it uses the default (`/tmp/nvimsocket` and `$NVIM_LISTEN_ADDRESS`). See `nvr --help`. Each neovim instance has a servername that can be identified by `:echo v:servername`. So, for the Skim backward search to work, one needs to ensure that Skim knows the servername of the correct neovim instance.

I now realize that the docs are confusing in this sense, or perhaps even plain wrong. I've updated the docs now. I removed the faq entry and merged it with the `vimtex-view-skim` entry. Please let me know what you think, and feel free to propose an update with a complete setup.

**Comment by lervag** [comment]
@clason I'm curios, what do you use for the backward search config in Skim? Do you use the simple `nvr --remote...`, or something else?

**Comment by clason** [comment]
Yes, that's right. That what splits were invented for, after all :)

**Comment by lervag** [comment]
I partly agree; I usually avoid having multiple things going on at the same time. But I'm not at all surprised that some people are very different, and I would want VimTeX to work as expected in most cases.

I believe the current docs more or less explain things well (see `:help vimtex-view-skim`). I'm currently referring to @jdhao's blog post, but perhaps the configuration could be listed explicitly in the vimtex docs instead?

**Comment by lervag** [comment]
I've pushed a large doc update now. I would be very happy if someone could read the updated texts, and these in particular:

* `:help vimtex-faq-neovim`
* `:help vimtex-view-skim`
* `:help vimtex-view-sumatrapdf`
* `:help vimtex-synctex-inverse-search`

**Comment by lervag** [comment]
> Thanks, great update)

Glad you like it!

> I don't know that much about vimtex ecosystem, but I checked your update for inverse search, and added a few comments to the commit ...

Thank you for the good suggestions/comments! Regarding [this comment](https://github.com/lervag/vimtex/commit/c3fd5fef5a4c9629d1b42e1d7d5983f242a3bc34#r58420027), i.e.:

> ..., but the previous ones (`nvr --remote-silent +%line '%file'`, `nvr --remote-expr vimtex#view#reverse_goto(%line, '%file')` , `nvr --listen /tmp/mynvimservername`) still doesn't

I don't quite get what you mean. Can you be more specific? I.e., _what_ does not work with these? Please notice that the "simple" solutions require either that one only has a single instance of neovim (i.e. servername is "trivial"), or that one uses the `--listen servername` argument and provide this specific servername to the Skim settings. Only the solution by @jdhao is more "dynamic" here.

@clason I remember someone saying it may be better to use a different option than `--remote-silent`, but I don't remember who. Was it you?

**Comment by clason** [comment]
> @clason I remember someone saying it may be better to use a different option than --remote-silent, but I don't remember who. Was it you?

I don't think so :)

> I am using only one instance, but probably I don't understand how to make them work.

Inverse search only works from the Skim instance started by Vimtex (for me) -- try closing Skim and starting via `<localleader>lv`.

**Comment by AlexFreik** [comment]
> I agree, but I'm not sure exactly how to help much more than by writing good docs. Suggestions are welcome, both on possible improvements/features and to improving the docs.

Maybe it is better to make Jhao's solution the main one for Mac/Linux (don't know about the Windows situation)? It is not that complex to add a few lines in init.vim, but could save time in the future and reduce issues like this one

**Comment by clason** [comment]
Ah, OK, I've only ever seen the issue raised in the context of `nvr` :D 

Then a neovism is of course less than ideal.

**Comment by lervag** [comment]
Yes, I also made the same conclusion first. But when I was rewriting all of this, I realized that we have the same exact challenges. I believe it may be somewhat more robust from Vim in that `vim --remote ...` will immediately use an existing server regardless of environment variables or whatever; if it exists. `nvr --remote ...` will use the default server `/tmp/nvimsocket`, which is basically never the correct servername. Thus, `nvr` works if the `NVIM_LISTEN_ADDRESS` exists or if you provide the correct servername explicitly.
